spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpNioConnection.java

```diff
@@ -139,6 +139,14 @@ private void doClose() {
 		}
 		catch (@SuppressWarnings(UNUSED) Exception e) {
 		}
+		try {
+			// in order to prevent endless recursive calls from channelOutputStream.close() to doClose() that would call channelOutputStream.close() etc...
+			if (this.channelOutputStream != null) {
+				((OutputStream) this.channelOutputStream).close();
```

Comment:
I don't understand how does your code help?
The `this.channelOutputStream` is never `null` according `TcpNioConnection` and you cannot set it to `null` (and you don't) because it is `final`.

And it is pointless to cast it to `OutputStream` since it is a super of the `ChannelOutputStream`.

Our idea was really to call this close here, but remove that `doClose()` from the `ChannelOutputStream`.

At the moment I'm very confused since the solution does not reflect your comments and does not address our discussion.

Please, elaborate what am I still missing?

Thanks

Reply:
I took a look again in the code the source of the trouble as it looked to me is the ```ChannelOutputStream.close()```
it is:
```@Override
		public void close() {
			doClose();
		}```
```doClose``` is not part of class ```ChannelOutputStream``` and what it does has nothing to do with ```ChannelOutputStream```, it nor handle the ```selector``` that is private member of ```ChannelOutputStream``` neither closes the OutputStream.
So I think that the best approach would be to rewrite the ```ChannelOutputStream.close()``` to:
```@Override
		public void close() {
			if (this.selector != null) {
				try {
					this.selector.close();
				}
				catch (IOException e) {
					// do nothing
				}
			}
			super.close();
		}```


I would wait to hear from you before commiting anything.
10x,
David